File: bitcoin/src/main/java/com/brentpanther/bitcoinwidget/WidgetApplication.java
package com.brentpanther.bitcoinwidget;

import android.app.Application;
import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.os.Build;
import android.preference.PreferenceManager;


public class WidgetApplication extends Application {

    private static WidgetApplication instance;

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;
        registerReceiver(new MyBroadcastReceiver(), new IntentFilter(Intent.ACTION_CONFIGURATION_CHANGED));
    }

    public static WidgetApplication getInstance() {
        return instance;
    }

    public int[] getWidgetIds() {
        AppWidgetManager manager = AppWidgetManager.getInstance(this);
        ComponentName cm = new ComponentName(this, WidgetProvider.class);
        return manager.getAppWidgetIds(cm);
    }

    public boolean useAutoSizing() {
        SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this);
        boolean fixedSize = sharedPrefs.getBoolean(getString(R.string.key_fixed_size), false);
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && !fixedSize;
    }

}

File: bitcoin/src/main/java/com/brentpanther/bitcoinwidget/WidgetApplication.kt
package com.brentpanther.bitcoinwidget

import android.app.Application
import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Intent
import android.content.IntentFilter
import android.content.res.Resources
import com.brentpanther.bitcoinwidget.receiver.WidgetBroadcastReceiver

class WidgetApplication : Application() {

    fun <T : WidgetProvider> getWidgetIds(className: Class<T>): IntArray {
        val name = ComponentName(this, className)
        return AppWidgetManager.getInstance(this).getAppWidgetIds(name)
    }
    val widgetIds: IntArray
        get() {
            return widgetProviders.map { getWidgetIds(it).toList() }.flatten().toIntArray()
        }

    val widgetProviders = listOf(WidgetProvider::class.java, ValueWidgetProvider::class.java)

    fun getWidgetType(widgetId: Int): WidgetType {
        val widgetInfo = AppWidgetManager.getInstance(this).getAppWidgetInfo(widgetId)
            ?: return WidgetType.PRICE
        return when (widgetInfo.provider.className) {
            WidgetProvider::class.qualifiedName -> WidgetType.PRICE
            ValueWidgetProvider::class.qualifiedName -> WidgetType.VALUE
            else -> throw IllegalArgumentException()
        }
    }

    override fun onCreate() {
        super.onCreate()
        instance = this
        registerReceiver(WidgetBroadcastReceiver(), IntentFilter(Intent.ACTION_CONFIGURATION_CHANGED))
        WidgetUpdater.updateDisplays(this)
    }

    companion object {

        lateinit var instance: WidgetApplication
            private set

        fun Int.dpToPx() = this * Resources.getSystem().displayMetrics.density
    }

}